File: frontend/src/types/signal.ts
export type SignalType =
  | 'ma_crossover_bullish'
  | 'ma_crossover_bearish'
  | 'rsi_oversold'
  | 'rsi_overbought'
  | 'macd_bullish_cross'
  | 'macd_bearish_cross'
  | 'near_52w_high'
  | 'near_52w_low'
  | 'new_52w_high'
  | 'new_52w_low';

export interface Signal {
  id: string;
  symbol: string;
  signal_type: SignalType;
  timestamp: string;
  price: number;
  details: Record<string, unknown>;
  created_at: string;
}

export interface SignalsResponse {
  signals: Signal[];
  total: number;
  filters: {
    types: string[] | null;
    symbols: string[] | null;
    hours: number;
  };
}

export interface SignalTypeInfo {
  type: SignalType;
  name: string;
  description: string;
  sentiment: 'bullish' | 'bearish' | 'neutral';
}

export const SIGNAL_TYPE_INFO: Record<SignalType, SignalTypeInfo> = {
  ma_crossover_bullish: {
    type: 'ma_crossover_bullish',
    name: 'MA Crossover Bullish',
    description: 'Price crosses above a moving average',
    sentiment: 'bullish',
  },
  ma_crossover_bearish: {
    type: 'ma_crossover_bearish',
    name: 'MA Crossover Bearish',
    description: 'Price crosses below a moving average',
    sentiment: 'bearish',
  },
  rsi_oversold: {
    type: 'rsi_oversold',
    name: 'RSI Oversold',
    description: 'RSI drops below 30',
    sentiment: 'bullish',
  },
  rsi_overbought: {
    type: 'rsi_overbought',
    name: 'RSI Overbought',
    description: 'RSI rises above 70',
    sentiment: 'bearish',
  },
  macd_bullish_cross: {
    type: 'macd_bullish_cross',
    name: 'MACD Bullish Cross',
    description: 'MACD line crosses above signal line',
    sentiment: 'bullish',
  },
  macd_bearish_cross: {
    type: 'macd_bearish_cross',
    name: 'MACD Bearish Cross',
    description: 'MACD line crosses below signal line',
    sentiment: 'bearish',
  },
  near_52w_high: {
    type: 'near_52w_high',
    name: 'Near 52W High',
    description: 'Within 5% of 52-week high',
    sentiment: 'neutral',
  },
  near_52w_low: {
    type: 'near_52w_low',
    name: 'Near 52W Low',
    description: 'Within 5% of 52-week low',
    sentiment: 'neutral',
  },
  new_52w_high: {
    type: 'new_52w_high',
    name: 'New 52W High',
    description: 'New 52-week high',
    sentiment: 'bullish',
  },
  new_52w_low: {
    type: 'new_52w_low',
    name: 'New 52W Low',
    description: 'New 52-week low',
    sentiment: 'bearish',
  },
};

File: backend/app/models/signal.py
"""Signal-related Pydantic models."""

from datetime import datetime, timezone
from enum import Enum
from typing import Any
from uuid import UUID, uuid4

from pydantic import BaseModel, Field


class SignalType(str, Enum):
    """Types of trading signals."""

    MA_CROSSOVER_BULLISH = "ma_crossover_bullish"
    MA_CROSSOVER_BEARISH = "ma_crossover_bearish"
    RSI_OVERSOLD = "rsi_oversold"
    RSI_OVERBOUGHT = "rsi_overbought"
    MACD_BULLISH_CROSS = "macd_bullish_cross"
    MACD_BEARISH_CROSS = "macd_bearish_cross"
    NEAR_52W_HIGH = "near_52w_high"
    NEAR_52W_LOW = "near_52w_low"
    NEW_52W_HIGH = "new_52w_high"
    NEW_52W_LOW = "new_52w_low"


class SignalCreate(BaseModel):
    """Schema for creating a new signal."""

    symbol: str
    signal_type: SignalType
    price: float
    details: dict[str, Any] = Field(default_factory=dict)


class Signal(BaseModel):
    """Complete signal with metadata."""

    id: UUID = Field(default_factory=uuid4)
    symbol: str
    signal_type: SignalType
    timestamp: datetime
    price: float
    details: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

    @property
    def is_bullish(self) -> bool:
        """Check if signal is bullish."""
        return self.signal_type in {
            SignalType.MA_CROSSOVER_BULLISH,
            SignalType.MACD_BULLISH_CROSS,
            SignalType.RSI_OVERSOLD,
            SignalType.NEAR_52W_LOW,
            SignalType.NEW_52W_HIGH,
        }

    @property
    def is_bearish(self) -> bool:
        """Check if signal is bearish."""
        return self.signal_type in {
            SignalType.MA_CROSSOVER_BEARISH,
            SignalType.MACD_BEARISH_CROSS,
            SignalType.RSI_OVERBOUGHT,
            SignalType.NEAR_52W_HIGH,
            SignalType.NEW_52W_LOW,
        }
